src/librbd/CopyupRequest.cc

```diff
@@ -146,7 +146,7 @@ bool CopyupRequest::send_copyup() {
     ldout(m_ictx->cct, 20) << __func__ << " " << this << " copyup with "
                            << "empty snapshot context" << dendl;
     librados::AioCompletion *comp = util::create_rados_safe_callback(this);
-    r = m_ictx->md_ctx.aio_operate(m_oid, comp, &copyup_op, 0, snaps);
+    r = m_ictx->data_ctx.aio_operate(m_oid, comp, &copyup_op, 0, snaps);
```

Comment:
@dillaman I understand what you mean and see why it's required. I tried to see what's happens   in the OSD when `m_ictx->data_ctx` is used and it seems that the `SnapContext` is empty in this case too (empty `snaps` during `rbd.copyup` and list of `SnapContext` for the actual (write) call):

```
2016-11-15 14:26:44.078037 7fc8807cc700 10 osd.2 pg_epoch: 44 pg[3.e( v 32'11 (0'0,32'11] local-les=35 n=5 ec=11 les/c/f 35/35/0 34/34/34) [2,0,1] r=0 lpr=34 crt=32'11 lcod 0'0 mlcod 0'0 active+clean] execute_ctx 0x560cf27f5c00
2016-11-15 14:26:44.078049 7fc8807cc700 10 osd.2 pg_epoch: 44 pg[3.e( v 32'11 (0'0,32'11] local-les=35 n=5 ec=11 les/c/f 35/35/0 34/34/34) [2,0,1] r=0 lpr=34 crt=32'11 lcod 0'0 mlcod 0'0 active+clean] do_op 3:71a2fc68:::rbd_data.85382ae8944a.0000000000000000:head [call rbd.copyup] ov 0'0 av 44'12 snapc 0=[] snapset 0=[]:[]

2016-11-15 14:26:44.220713 7fc8807cc700 10 osd.2 pg_epoch: 44 pg[3.e( v 44'12 (0'0,44'12] local-les=35 n=6 ec=11 les/c/f 35/35/0 34/34/34) [2,0,1] r=0 lpr=34 crt=32'11 lcod 32'11 mlcod 32'11 active+clean] execute_ctx 0x560cf2a8f500
2016-11-15 14:26:44.220726 7fc8807cc700 10 osd.2 pg_epoch: 44 pg[3.e( v 44'12 (0'0,44'12] local-les=35 n=6 ec=11 les/c/f 35/35/0 34/34/34) [2,0,1] r=0 lpr=34 crt=32'11 lcod 32'11 mlcod 32'11 active+clean] do_op 3:71a2fc68:::rbd_data.85382ae8944a.0000000000000000:head [set-alloc-hint object_size 4194304 write_size 4194304,write 0~0] ov 44'12 av 44'13 snapc 14=[14,13,12,11,10,f,e,d] snapset 0=[]:[]+head
```

Note that I've made the changes as suggested though.